src/blr/cli/tagfastq.py

```diff
@@ -119,22 +101,44 @@ def parse_corrected_barcodes(open_file):
     return corrected_barcodes
 
 
-def parse_raw_barcodes(open_file):
-    """
-    Parse FASTA/FASTQ containing barcode sequences and return a dictionary with entry headers
-    pointing to a raw barcodes sequence.
-    :param open_file: dnaio odject.
-    :return: dict: entry headers pointing to a raw barcodes sequence
-    """
-    for barcode in tqdm(open_file, desc="Raw barcodes processed"):
-        header, _ = barcode.name.split(maxsplit=1)
-        yield {header: barcode.sequence}
+class BarcodeReader:
+    def __init__(self, filename):
+        self._cache = dict()
+        self._file = dnaio.open(filename, mode="r")
+        self.barcodes = self.parse()
+
+    def parse(self):
+        for barcode in tqdm(self._file, desc="Uncorrected barcodes processed"):
+            header, _ = barcode.name.split(maxsplit=1)
+            yield {header: barcode.sequence}
+
+    def get_barcode(self, header, maxiter=10):
+        if header in self._cache:
+            return self._cache.pop(header)
+
+        for header_barcode_pair in islice(self.barcodes, maxiter):
+            # If header in next pair then parser lines are synced --> drop cache.
+            if header in header_barcode_pair:
+                self._cache.clear()
+                return header_barcode_pair[header]
+
+            self._cache.update(header_barcode_pair)
+        return None
```

Comment:
I think this could be written with less return statements. I think it's easier to have an overview if there is only one return statement.

```suggestion
    def get_barcode(self, header, maxiter=10):
        if header in self._cache:
            barcode = self._cache.pop(header)
        else:
            for header_barcode_pair in islice(self.barcodes, maxiter):
                # If header in next pair then parser lines are synced --> drop cache.
                if header in header_barcode_pair:
                    self._cache.clear()
                    barcode = header_barcode_pair[header]
                self._cache.update(header_barcode_pair)
        return barcode
```

Reply:
I actually prefer mine with the manny return statements. This avoids assigning a value to the `barcode` variable before returning it.  Mine is also less complex in my opinion with less indentation levels. 

Note also add that in order to  have the same functionality as my code the rewrite would have to look a bit more like this:

```python
def get_barcode(self, header, maxiter=10):
   if header in self._cache:
      barcode = self._cache.pop(header)
   else:
      barcode = None
      for header_barcode_pair in islice(self.barcodes, maxiter):
         # If header in next pair then parser lines are synced --> drop cache.
         if header in header_barcode_pair:
            self._cache.clear()
            barcode = header_barcode_pair[header]
            break
         self._cache.update(header_barcode_pair)
   return barcode
```
The `barcode = None` needs to be included in the case that the barcode is not present. The `break` should be included if you want to stop parsing if the barcode is found. If not included the cache will be filled 10 barcodes at the time including any matched barcode. 

I tested running 10 million reads using your code without the `break` and it works fine. Performance similar to my code.

Reply:
I also think @pontushojer’s version is fine, but for completeness, initializing `barcode` to `None` could be done by using a default value for `pop`:
```
   barcode = self._cache.pop(header, None)
   if barcode is None:
```

Reply:
Alright